lib/internal/test_runner/runner.js

```diff
@@ -170,17 +184,46 @@ function run(options) {
   if (options === null || typeof options !== 'object') {
     options = kEmptyObject;
   }
-  const { concurrency, timeout, signal, files, inspectPort } = options;
+  const { concurrency, timeout, signal, files, inspectPort, watch } = options;
 
   if (files != null) {
     validateArray(files, 'options.files');
   }
+  if (watch != null) {
+    validateBoolean(watch, 'options.watch');
+  }
 
   const root = createTestTree({ concurrency, timeout, signal });
   const testFiles = files ?? createTestFileList();
 
-  PromisePrototypeThen(SafePromiseAll(testFiles, (path) => runTestFile(path, root, inspectPort)),
-                       () => root.postRun());
+  let postRun = () => root.postRun();
+  let filesWatcher;
+  const runningSubtests = new SafeMap();
+  if (watch) {
+    filesWatcher = new FilesWatcher({ throttle: 500, mode: 'filter' });
+    filesWatcher.on('changed', ({ owners }) => {
+      filesWatcher.unfilterFilesOwnedBy(owners);
+      const dependantTests = ArrayPrototypeFilter(testFiles, (file) => owners.has(file));
+      ArrayPrototypeForEach(dependantTests, async (file) => {
+        const runningProcess = runningProcesses.get(file);
+        if (runningProcess) {
+          runningProcess.kill();
+          await once(runningProcess, 'exit');
+        }
+        await runningSubtests.get(file);
+        runningSubtests.set(file, runTestFile(file, root, inspectPort, filesWatcher));
+      });
```

Comment:
Can you add a comment somewhere explaining why we create those async function without awaiting them / without attaching `.catch` handlers (and we probably should btw)